src/modules/dotnet.rs

```diff
@@ -0,0 +1,308 @@
+                            buffer.push_str(version_string);
+                            Some(Version(buffer))
+                        }
+                        _ => None,
+                    }
+                }
+                _ => None,
+            }
+        }
+        _ => None,
+    }
+}
+
+fn get_local_dotnet_files<'a>(context: &'a Context) -> Result<Vec<DotNetFile<'a>>, std::io::Error> {
+    Ok(context
+        .get_dir_files()?
+        .iter()
+        .filter_map(|p| {
+            get_dotnet_file_type(p).map(|t| DotNetFile {
+                path: p.as_ref(),
+                file_type: t,
+            })
+        })
+        .collect())
+}
+
+fn get_dotnet_file_type(path: &Path) -> Option<FileType> {
+    let file_name_lower = map_str_to_lower(path.file_name());
+
+    match file_name_lower.as_ref().map(|f| f.as_ref()) {
+        Some(GLOBAL_JSON_FILE) => return Some(FileType::GlobalJson),
+        Some(PROJECT_JSON_FILE) => return Some(FileType::ProjectJson),
+        _ => (),
+    };
+
+    let extension_lower = map_str_to_lower(path.extension());
+
+    match extension_lower.as_ref().map(|f| f.as_ref()) {
+        Some("sln") => return Some(FileType::SolutionFile),
+        Some("csproj") | Some("fsproj") | Some("xproj") => return Some(FileType::ProjectFile),
+        _ => (),
+    };
+
+    None
+}
+
+fn map_str_to_lower(value: Option<&OsStr>) -> Option<String> {
+    Some(value?.to_str()?.to_ascii_lowercase())
+}
+
+fn get_version_from_cli() -> Option<Version> {
+    let version_output = match Command::new("dotnet").arg("--version").output() {
+        Ok(output) => output,
+        Err(e) => {
+            log::warn!("Failed to execute `dotnet --version`. {}", e);
+            return None;
+        }
+    };
+    let version = str::from_utf8(version_output.stdout.as_slice())
+        .ok()?
+        .trim();
+
+    let mut buffer = String::with_capacity(version.len() + 1);
+    buffer.push('v');
+    buffer.push_str(version);
+
+    Some(Version(buffer))
+}
+
+fn get_latest_sdk_from_cli() -> Option<Version> {
+    let mut cmd = Command::new("dotnet");
+    cmd.arg("--list-sdks")
+        .stdout(Stdio::piped())
+        .stderr(Stdio::null())
+        .stdin(Stdio::null());
+
+    let exit_code = match cmd.status() {
+        Ok(status) => status,
+        Err(e) => {
+            log::warn!("Failed to execute `dotnet --list-sdks`. {}", e);
+            return None;
+        }
+    };
+
+    if exit_code.success() {
+        let sdks_output = cmd.output().ok()?;
+        fn parse_failed<T>() -> Option<T> {
+            log::warn!("Unable to parse the output from `dotnet --list-sdks`.");
+            None
+        };
+        let latest_sdk = str::from_utf8(sdks_output.stdout.as_slice())
+            .ok()?
+            .lines()
+            .map(str::trim)
+            .filter(|l| !l.is_empty())
+            .last()
+            .or_else(parse_failed)?;
+        let take_until = latest_sdk.find('[')? - 1;
```

Comment:
While this is going to be a pain in terms of code length, would it make sense to log a statement here (e.g. using parse_failed) to let the user know that their dotnet isn't playing nice?

Reply:
`parse_failed` already logs a warning.. Or are you referring to some different output?

Reply:
Done! 😄